discovery/kubernetes/endpoints.go

```diff
@@ -152,31 +110,85 @@ func (e *Endpoints) Run(ctx context.Context, ch chan<- []*targetgroup.Group) {
 		},
 	})
 
+	return e
+}
+
+func (e *Endpoints) enqueue(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		return
+	}
+
+	e.queue.Add(key)
+}
+
+// Run implements the Discoverer interface.
+func (e *Endpoints) Run(ctx context.Context, ch chan<- []*targetgroup.Group) {
+	defer e.queue.ShutDown()
+
+	if !cache.WaitForCacheSync(ctx.Done(), e.endpointsInf.HasSynced, e.serviceInf.HasSynced, e.podInf.HasSynced) {
+		level.Error(e.logger).Log("msg", "endpoints informer unable to sync cache")
+		return
+	}
+
+	go func() {
+		for e.process(ctx, ch) {
+		}
+	}()
+
 	// Block until the target provider is explicitly canceled.
 	<-ctx.Done()
 }
 
+func (e *Endpoints) process(ctx context.Context, ch chan<- []*targetgroup.Group) bool {
+	keyObj, quit := e.queue.Get()
+	if quit {
+		return false
+	}
+	defer e.queue.Done(keyObj)
+	key := keyObj.(string)
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		level.Error(e.logger).Log("msg", "spliting key failed", "key", key)
+		return true
+	}
+
+	o, exists, err := e.endpointsStore.GetByKey(key)
+	if err != nil {
+		level.Error(e.logger).Log("msg", "getting object from store failed", "key", key)
+		return true
+	}
+	if !exists {
+		send(ctx, e.logger, RoleEndpoint, ch, &targetgroup.Group{Source: endpointsSourceFromNamespaceAndName(namespace, name)})
+		return true
+	}
+	eps, err := convertToEndpoints(o)
+	if err != nil {
+		level.Error(e.logger).Log("msg", "converting to Endpoints object failed", "err", err)
+		return true
+	}
+	send(ctx, e.logger, RoleEndpoint, ch, e.buildEndpoints(eps))
+	return true
+}
+
 func convertToEndpoints(o interface{}) (*apiv1.Endpoints, error) {
 	endpoints, ok := o.(*apiv1.Endpoints)
 	if ok {
 		return endpoints, nil
 	}
 
-	deletedState, ok := o.(cache.DeletedFinalStateUnknown)
-	if !ok {
-		return nil, fmt.Errorf("Received unexpected object: %v", o)
-	}
-	endpoints, ok = deletedState.Obj.(*apiv1.Endpoints)
-	if !ok {
-		return nil, fmt.Errorf("DeletedFinalStateUnknown contained non-Endpoints object: %v", deletedState.Obj)
-	}
-	return endpoints, nil
+	return nil, fmt.Errorf("Received unexpected object: %v", o)
 }
 
 func endpointsSource(ep *apiv1.Endpoints) string {
```

Comment:
nit, it could be rewritten like this to avoid duplication:

```go
func endpointsSource(ep *apiv1.Endpoints) string {
    return endpointsSourceFromNamespaceAndName(s.Namespace, s.Name)
}
```

The same comment applies for the `<role>Source(...)` functions in the other files.